transformer_engine/pytorch/optimizers/fused_adam.py

```diff
@@ -73,6 +99,8 @@ def __init__(
         set_grad_none=True,
         capturable=False,
         master_weights=False,
+        extra_param_groups=None,
```

Comment:
Aren't the `master_weights` redundant with `extra_param_groups`? In FP16 training, the model weights are FP16 and the master weights are FP32. For FP8 training it would be intuitive to have model weights in `Float8Tensor` and master weights in FP32.

Having a consistent API between FP16 and FP8 training will involve some tweaks to the FP8 kernel: https://github.com/NVIDIA/TransformerEngine/blob/c71e5e989ee09078249cfaf5ab27f24db059a329/transformer_engine/pytorch/csrc/extensions/multi_tensor/multi_tensor_adam.cu#L153

Reply:
Yeah. This part confused me for quite some time as well.
FusedAdam has the capability to simultaneously update both the master weights (fp32) and model weights (fp16/bf16) by setting master_weights=True. However, I've noticed that Megatron-LM never utilize this feature. Instead, it sets master_weights=False and capturable=False always. Megatron-LM provides FusedAdam with param_groups containing only the FP32 master weights from the `DistributedOptimizer`. FusedAdam then treats these as model weights, updating them in-place. Then it uses `copy_main_params_to_model_params` to copy these master weights to the model weights in the dist opt.
I was hesitant to remove the master_weights=False option in FusedAdam to avoid potentially disrupting existing functionality, it does appear to be redundant in this context.

Reply:
Also, `master_weights=True` is designed to be used together with `capturable=True`. I am not sure where these features are being used. https://github.com/NVIDIA/TransformerEngine/blob/38b48c8abd6f03c284a6f0864186865d66f1e8f5/transformer_engine/pytorch/optimizers/fused_adam.py#L80-L83

Reply:
I'm uncomfortable with having both `master_weights` and `extra_param_groups`, since they accomplish the same things but are backwards from each other. This will definitely cause confusion and bugs in the future. This PR has the kernels to fully support `master_weights=True` so I think the cleanest approach is to commit to that API and adapt Megatron-LM to use it.

Reply:
Yeah, you are right. I need some time to refactor and test with Megatron-LM.

Reply:
Hi @timmoon10 , I am trying to modify Megatron-LM to accommodate FusedAdam's design, but using Float8Tensor as model weights in FusedAdam (fp8_params, lr, ...) leads to some strange errors. Additionally, the `master_weights` and `capturable` features in FusedAdam are currently tightly coupled, and we don't need the capturable functionality at the moment.
Therefore, I have to continue treating Megatron's DistributedOptimizer's master weights as the model weights for FusedAdam. I've removed the `extra_param_groups` argument and refactored the `step` function, which should make the code more comprehensible.

Reply:
I think it's more confusing that `master_weights` has opposite meanings with `capturable=True` (master weights are read-write, model weights are write-only) and `capturable=False` (master weights are write-only, model weights are read-write).

I see a few options:
1. Settle on the `master_weights` API. Need to modify Megatron-LM so the optimizer can hold `Float8Tensor` params.
2. Settle on the `extra_weights` API. Need to modify the capturable kernels so it treats the optimizer's params as master params.
3. Have both `master_weights` and `extra_weights`, which are used for different kernels.
4. Remove support for capturable kernels.
5. Use `master_weights` inconsistently depending on the kernel.

I'd prefer 1 or 2, but it looks like 3 might be the most realistic.